<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from powerball.urdf.xacro           | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="powerball"  xmlns:xacro="http://ros.org/wiki/xacro">
  <!-- common stuff -->
  <material name="Blue">
    <color rgba="0.0 0.0 0.8 1.0"/>
  </material>
  <material name="Green">
    <color rgba="0.0 0.8 0.0 1.0"/>
  </material>
  <material name="Grey">
    <color rgba="0.7 0.7 0.7 "/>
  </material>
  <material name="Grey2">
    <color rgba="0.9 0.9 0.9 1.0"/>
  </material>
  <material name="Red">
    <color rgba="0.8 0.0 0.0 1.0"/>
  </material>
  <material name="White">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>
  <material name="Black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="Caster">
    <texture filename="package://pr2_description/materials/textures/pr2_caster_texture.png"/>
  </material>
  <material name="Wheel_l">
    <texture filename="package://pr2_description/materials/textures/pr2_wheel_left.png"/>
  </material>
  <material name="Wheel_r">
    <texture filename="package://pr2_description/materials/textures/pr2_wheel_right.png"/>
  </material>
  <material name="RollLinks">
    <texture filename="package://pr2_description/materials/textures/pr2_wheel_left.png"/>
  </material>
     <!-- GazeboMechanismControl -->
    <controller:gazebo_ros_controller_manager name="gazebo_ros_controller_manager" plugin="libgazebo_ros_controller_manager.so">
      <alwaysOn>true</alwaysOn>
      <updateRate>1000.0</updateRate>
      <robotParam>/arm_controller/state</robotParam>
      <robotNamespace>/</robotNamespace>
    </controller:gazebo_ros_controller_manager> 
       <controller:gazebo_ros_power_monitor name="gazebo_ros_power_monitor_controller" plugin="libgazebo_ros_power_monitor.so">
        <alwaysOn>true</alwaysOn>
        <updateRate>1.0</updateRate>
        <timeout>5</timeout>
        <interface:audio name="power_monitor_dummy_interface" />
        <powerStateTopic>power_state</powerStateTopic>
        <powerStateRate>10.0</powerStateRate>
        <fullChargeCapacity>87.78</fullChargeCapacity>
        <dischargeRate>-474</dischargeRate>
        <chargeRate>525</chargeRate>
        <dischargeVoltage>15.52</dischargeVoltage>
        <chargeVoltage>16.41</chargeVoltage>
    </controller:gazebo_ros_power_monitor>
    <!--
      <alwaysOn>true</alwaysOn>
      <updateRate>1000.0</updateRate>
      <interface:audio name="gazebo_ros_controller_manager_dummy_iface"/>
    </controller:gazebo_ros_controller_manager>
    <controller:gazebo_ros_power_monitor name="gazebo_ros_power_monitor_controller" plugin="libgazebo_ros_power_monitor.so">
      <alwaysOn>true</alwaysOn>
      <updateRate>1.0</updateRate>
      <timeout>5</timeout>
      <interface:audio name="power_monitor_dummy_interface"/>
      <powerStateTopic>power_state</powerStateTopic>
      <powerStateRate>10.0</powerStateRate>
      <fullChargeCapacity>87.78</fullChargeCapacity>
      <dischargeRate>-474</dischargeRate>
      <chargeRate>525</chargeRate>
      <dischargeVoltage>15.52</dischargeVoltage>
      <chargeVoltage>16.41</chargeVoltage>
    </controller:gazebo_ros_power_monitor>
  </gazebo> -->
  <!-- calibration -->
  <!-- schunk powerball -->
  <!-- foot for arm -->
  <link name="base_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 -10"/>
      <mass value="1000.0"/>
      <inertia ixx="100.0" ixy="0" ixz="0" iyy="100.0" iyz="0" izz="100.0"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.25"/>
      <geometry>
        <cylinder length="0.5" radius="0.1"/>
      </geometry>
      <material name="Grey1"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.25"/>
      <geometry>
        <cylinder length="0.5" radius="0.1"/>
      </geometry>
    </collision>
  </link>
  <!-- arm -->
  <joint name="arm_0_joint" type="fixed">
    <origin rpy="3.1415926 0 0" xyz="0 0 0.5"/>
    <parent link="base_link"/>
    <child link="arm_0_link"/>
  </joint>
  <link name="arm_0_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.29364"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball0.stl"/>
      </geometry>
      <material name="Grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball0.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="arm_1_joint" type="revolute">
    <origin rpy="0 0 0" xyz="0 0 -0.0897"/>
    <parent link="arm_0_link"/>
    <child link="arm_1_link"/>
    <axis xyz="0 0 -1"/>
    <!--axis xyz="0 0 1"/-->
    <calibration rising="0"/>
    <dynamics damping="10"/>
    <limit effort="370" lower="-2.97" upper="2.97" velocity="2.0"/>
    <safety_controller k_position="20" k_velocity="50" soft_lower_limit="-6.2731853" soft_upper_limit="6.2731853"/>
  </joint>
  <link name="arm_1_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.29364"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball1.stl"/>
      </geometry>
      <material name="Blue"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball1.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="arm_2_joint" type="revolute">
    <origin rpy="1.5708 0 0" xyz="0 0 0"/>
    <parent link="arm_1_link"/>
    <child link="arm_2_link"/>
    <axis xyz="0 0 -1"/>
    <calibration rising="0"/>
    <dynamics damping="10"/>
    <limit effort="370" lower="-1.92" upper="1.92" velocity="2.0"/>
    <safety_controller k_position="20" k_velocity="50" soft_lower_limit="-2.0843" soft_upper_limit="2.0843"/>
  </joint>
  <link name="arm_2_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1.68311"/>
      <inertia ixx="0.03" ixy="0" ixz="0" iyy="0.03" iyz="0" izz="0.03"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball2.stl"/>
      </geometry>
      <material name="Grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball2.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="arm_3_joint" type="revolute">
    <origin rpy="3.14159 0 0" xyz="0 -0.350 0"/>
    <parent link="arm_2_link"/>
    <child link="arm_3_link"/>
    <axis xyz="0 0 -1"/>
    <calibration rising="0"/>
    <dynamics damping="5"/>
    <limit effort="176" lower="-2.71" upper="2.71" velocity="2.0"/>
    <safety_controller k_position="20" k_velocity="25" soft_lower_limit="-6.2731853" soft_upper_limit="6.2731853"/>
  </joint>
  <link name="arm_3_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="2.1"/>
      <inertia ixx="0.03" ixy="0" ixz="0" iyy="0.03" iyz="0" izz="0.03"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball3.stl"/>
      </geometry>
      <material name="Blue"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball3.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="arm_4_joint" type="revolute">
    <origin rpy="1.5708 3.14159 0" xyz="0 0 0"/>
    <parent link="arm_3_link"/>
    <child link="arm_4_link"/>
    <axis xyz="0 0 1"/>
    <calibration rising="0"/>
    <dynamics damping="5"/>
    <limit effort="176" lower="-2.97" upper="2.97" velocity="2.0"/>
    <safety_controller k_position="20" k_velocity="25" soft_lower_limit="-2.0843951" soft_upper_limit="2.0843951"/>
  </joint>
  <link name="arm_4_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1.68311"/>
      <inertia ixx="0.03" ixy="0" ixz="0" iyy="0.03" iyz="0" izz="0.03"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball4.stl"/>
      </geometry>
      <material name="Grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball4.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="arm_5_joint" type="revolute">
    <origin rpy="1.5708 0 0" xyz="0 0.005 -0.3049"/>
    <parent link="arm_4_link"/>
    <child link="arm_5_link"/>
    <axis xyz="0 0 -1"/>
    <calibration rising="0"/>
    <dynamics damping="5"/>
    <limit effort="41.6" lower="-2.44" upper="2.44" velocity="2.0"/>
    <safety_controller k_position="20" k_velocity="25" soft_lower_limit="-6.2731853" soft_upper_limit="6.2731853"/>
  </joint>
  <link name="arm_5_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.807"/>
      <inertia ixx="0.03" ixy="0" ixz="0" iyy="0.03" iyz="0" izz="0.03"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball5.stl"/>
      </geometry>
      <material name="Blue"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball5.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="arm_6_joint" type="revolute">
    <origin rpy="-1.5708 0 0" xyz="0 0 0"/>
    <parent link="arm_5_link"/>
    <child link="arm_6_link"/>
    <axis xyz="0 0 -1"/>
    <calibration rising="0"/>
    <dynamics damping="5"/>
    <limit effort="20.1" lower="-2.97" upper="2.97" velocity="2.0"/>
    <safety_controller k_position="20" k_velocity="25" soft_lower_limit="-2.0843951" soft_upper_limit="2.0843951"/>
  </joint>
  <link name="arm_6_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.819"/>
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball6.stl"/>
      </geometry>
      <material name="Grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robotnik_powerball_description/meshes/powerball6.stl"/>
      </geometry>
    </collision>
  </link>
  <gazebo reference="arm_0_link">
    <material value="Grey"/>
    <turnGravityOff>false</turnGravityOff>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="arm_1_link">
    <material value="Blue"/>
    <turnGravityOff>false</turnGravityOff>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="arm_2_link">
    <material value="Grey"/>
    <turnGravityOff>false</turnGravityOff>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="arm_3_link">
    <material value="Blue"/>
    <turnGravityOff>false</turnGravityOff>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="arm_4_link">
    <material value="Grey"/>
    <turnGravityOff>false</turnGravityOff>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="arm_5_link">
    <material value="Blue"/>
    <turnGravityOff>false</turnGravityOff>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="arm_6_link">
    <material value="Grey"/>
    <turnGravityOff>false</turnGravityOff>
    <selfCollide>true</selfCollide>
  </gazebo>
  <transmission name="arm_1_trans" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="arm_1_motor"/>
    <joint name="arm_1_joint"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
  <transmission name="arm_2_trans" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="arm_2_motor"/>
    <joint name="arm_2_joint"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
  <transmission name="arm_3_trans" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="arm_3_motor"/>
    <joint name="arm_3_joint"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
  <transmission name="arm_4_trans" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="arm_4_motor"/>
    <joint name="arm_4_joint"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
  <transmission name="arm_5_trans" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="arm_5_motor"/>
    <joint name="arm_5_joint"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>
  <transmission name="arm_6_trans" type="pr2_mechanism_model/SimpleTransmission">
    <actuator name="arm_6_motor"/>
    <joint name="arm_6_joint"/>
    <mechanicalReduction>1</mechanicalReduction>
  </transmission>

</robot>

